File: tools/book.py
"""
book.py - 小说数据模型模块

定义 Book 类，负责从轻小说文库（wenku8.net）抓取并封装一部小说的完整信息：
  - 基本信息（标题、作者、状态、简介、封面等）
  - 章节目录（按卷分组）
  - 插图链接
  - 标题格式化工具
"""

import re
from typing import List, Optional

from .crawler import WebCrawler


class Book:
    """
    代表轻小说文库中的一部小说。

    实例化时会自动抓取该书的章节目录与基本信息，
    后续可通过各方法获取插图链接、封面内容等。
    """

    def __init__(self, book_id: str, crawler: WebCrawler = None):
        """
        初始化 Book 对象，自动拉取书籍数据。

        :param book_id: 书籍在文库中的数字编号，如 "3057"
        :param crawler: 可选，外部传入的 WebCrawler 实例；
                        若为 None，则在内部创建一个新实例
        """
        self.my_crawler = crawler if crawler is not None else WebCrawler()
        self.book_id = book_id

        # 按顺序初始化：章节目录 → 插图 URL 映射 → 基本信息
        self.base_chapter_url, self.volumes = self._fetch_chapters()
        self.volume_num = len(self.volumes)
        self.picture_urls = self._build_picture_url_map()
        self.basic_info = self._fetch_basic_info()

    # ------------------------------------------------------------------
    # 初始化阶段的私有抓取方法
    # ------------------------------------------------------------------

    def _fetch_chapters(self) -> tuple[str, dict]:
        """
        抓取小说的章节目录，并按卷分组。

        访问书籍详情页，找到「小说目录」链接，
        再抓取目录页，将章节解析为以卷名为键的有序字典。

        :return: (base_chapter_url, volumes) 元组
                 - base_chapter_url: 章节页的公共前缀 URL
                 - volumes: OrderedDict，键为卷名，值为章节列表
                            每个章节格式：{"name": str, "link": str}
        :raises Exception: 若找不到目录链接则抛出
        """
        book_page_url = f"https://www.wenku8.net/book/{self.book_id}.htm"
        soup = self.my_crawler.fetch(book_page_url)

        # 在详情页中找到「小说目录」的链接
        chapter_index_url = "https://www.wenku8.net"
        for a in soup.find("div", id="content").find("div").find_all("a"):
            if a.text == "小说目录":
                chapter_index_url += a["href"]
                break
        else:
            raise Exception(f"未找到小说目录链接，书籍编号：{self.book_id}")

        # 抓取目录页，解析卷和章节
        rows = (
            self.my_crawler.fetch(chapter_index_url)
            .find("table", class_="css")
            .find_all("tr")
        )

        volumes = {}
        current_volume = None

        for row in rows:
            # 检测卷标题行
            volume_tag = row.find("td", class_="vcss")
            if volume_tag:
                current_volume = volume_tag.get_text().strip()
                volumes[current_volume] = []
                continue

            # 检测章节行：提取所有带 href 的 <a> 标签
            link_tags = row.find_all("a", href=True)
            if link_tags and current_volume is not None:
                for tag in link_tags:
                    volumes[current_volume].append({
                        "name": tag.get_text().strip(),
                        "link": tag["href"],
                    })

        # 构建章节页的公共前缀（去掉 index.htm）
        base_chapter_url = chapter_index_url.replace("index.htm", "")
        return base_chapter_url, volumes

    def _build_picture_url_map(self) -> dict:
        """
        从章节目录中提取每一卷「插图」章节的链接，构建插图 URL 映射。

        :return: 字典，键为卷名，值为该卷插图章节的相对路径
                 如 {"第一卷": "125425.htm"}
        """
        picture_map = {}
        for volume_name, chapters in self.volumes.items():
            for chapter in chapters:
                if chapter["name"] == "插图":
                    picture_map[volume_name] = chapter["link"]
        return picture_map

    def _fetch_basic_info(self) -> dict:
        """
        抓取书籍详情页，解析基本信息并以字典形式返回。

        :return: 包含以下键的字典：
                 标题、作者、出版社、最新章节、连载状态、
                 更新时间、全文长度、简介、cover（封面 URL）
        """
        url = f"https://www.wenku8.net/book/{self.book_id}.htm"
        soup = self.my_crawler.fetch(url)
        content_div = soup.find("div", id="content")
        info_table = content_div.find("table")

        # 书名
        title = info_table.find("b").text

        # 基本属性（分类、作者、状态、更新时间、字数）
        info_cells = info_table.find_all("tr")[2].find_all("td")
        category = info_cells[0].text.split("：")[1]
        author = info_cells[1].text.split("：")[1]
        status = info_cells[2].text.split("：")[1]
        update_time = info_cells[3].text.split("：")[1] if len(info_cells) > 3 else None
        length = info_cells[4].text.split("：")[1] if len(info_cells) > 4 else None

        # 最新章节（部分书籍可能缺少该字段）
        try:
            latest_table = content_div.find("div").find_all("table")[2]
            latest_chapter = latest_table.find("a").get_text(strip=True)
        except Exception:
            latest_chapter = None

        # 简介（取最后一个 <span> 的文本）
        description = content_div.find_all("span")[-1].get_text()

        # 封面图（页面第二张 <img>，第一张是站点 Logo）
        img_urls = [img["src"] for img in soup.find_all("img") if "src" in img.attrs]
        cover = img_urls[1] if len(img_urls) > 1 else None

        return {
            "标题": title,
            "作者": author,
            "出版社": category,
            "最新章节": latest_chapter,
            "连载状态": status,
            "更新时间": update_time,
            "全文长度": length,
            "简介": description,
            "cover": cover,
        }

    # ------------------------------------------------------------------
    # 公开方法
    # ------------------------------------------------------------------

    def get_chapter_image_urls(self, volume_name: str = None) -> Optional[List[str]]:
        """
        获取指定卷插图章节页中所有图片的完整 URL。

        :param volume_name: 卷名；若为 None 或该卷无插图，则返回 None
        :return: 图片 URL 字符串列表，例如
                 ["https://pic.xxx.xyz/3/3057/125425/165682.jpg", ...]
        """
        if not volume_name or volume_name not in self.picture_urls:
            return None

        page_url = f"{self.base_chapter_url}{self.picture_urls[volume_name]}"
        img_tags = self.my_crawler.fetch(page_url).find_all("img")
        return [img["src"] for img in img_tags if "src" in img.attrs]

    def get_cover_content(self) -> bytes:
        """
        下载封面图片并返回原始二进制内容。

        :return: 封面图片的 bytes 内容
        """
        return self.my_crawler.fetch(self.basic_info["cover"], parse=False)

    def get_formatted_title(self, fmt: str) -> str:
        """
        根据指定格式返回书名。

        书名若包含括号（如"败北女角太多了！(败犬女主太多了！)"），
        可选择保留全名、仅保留括号外，或仅保留括号内。

        :param fmt: 格式代码，可选值：
                    - "FULL"：完整书名
                    - "OUT" ：括号外部分，如 "败北女角太多了！"
                    - "IN"  ：括号内部分，如 "败犬女主太多了！"
        :return: 格式化后的书名字符串；
                 若格式无法匹配（书名无括号时选 OUT/IN），返回原书名
        """
        title = self.basic_info["标题"]
        if fmt == "FULL":
            return title

        match = re.match(r"^(.*?)\((.*?)\)$", title)
        if match:
            if fmt == "OUT":
                return match.group(1)
            if fmt == "IN":
                return match.group(2)

        # 无法按格式解析时，回退到完整书名
        return title

File: src/main/book.ts
import type {
  CrawlerRequestControlFactory,
  CrawlerRequestKind,
  WebCrawler,
} from './crawler'
import type { TitleFormat } from '../shared/config-types'
import type {
  BasicInfo,
  BookVersionFields,
  Chapter,
} from '../shared/book-types'
import { formatBookTitle } from '../shared/title-format'
import {
  createBookVersion,
  type BookSnapshot,
  type BookVersion,
} from './book-cache-model'
import type { BookResourceCache } from './book-cache-repository'
import {
  DownloadCancelledError,
  throwIfDownloadCancelled,
} from './download-cancellation'

export interface ParsedBookPage {
  basicInfo: BasicInfo
  chapterIndexUrl: string
  versionFields: BookVersionFields
}

interface ChapterImageLoad {
  controller: AbortController
  promise: Promise<string[] | null>
  waiters: Set<symbol>
}

function waitForChapterImages(
  entry: ChapterImageLoad,
  signal?: AbortSignal,
): Promise<string[] | null> {
  throwIfDownloadCancelled(signal)
  const waiter = Symbol('chapter-image-waiter')
  entry.waiters.add(waiter)
  return new Promise((resolve, reject) => {
    let settled = false
    const cleanup = (): void => {
      if (settled) return
      settled = true
      signal?.removeEventListener('abort', onAbort)
      entry.waiters.delete(waiter)
      if (entry.waiters.size === 0 && !entry.controller.signal.aborted) {
        entry.controller.abort()
      }
    }
    const onAbort = (): void => {
      cleanup()
      reject(new DownloadCancelledError())
    }
    signal?.addEventListener('abort', onAbort, { once: true })
    if (signal?.aborted) {
      onAbort()
      return
    }
    entry.promise.then(
      (urls) => {
        cleanup()
        resolve(urls)
      },
      (error: unknown) => {
        cleanup()
        reject(error)
      },
    )
  })
}

export class Book {
  readonly bookId: string
  readonly version: BookVersion
  readonly legacyImportGenerationKey: string
  baseChapterUrl: string = ''
  volumes: Record<string, Chapter[]> = {}
  pictureUrls: Record<string, string> = {}
  basicInfo: BasicInfo = {
    '标题': '',
    '作者': '',
    '出版社': '',
    '最新章节': null,
    '连载状态': '',
    '更新时间': null,
    '全文长度': null,
    '简介': '',
    '标签': [],
    '动画化': false,
    '热度': null,
    'cover': null,
  }

  private readonly crawler: WebCrawler
  private readonly requestControlFactory?: CrawlerRequestControlFactory
  private readonly resources?: BookResourceCache
  private readonly chapterImageUrlCache = new Map<string, string[] | null>()
  private readonly chapterImageUrlLoads = new Map<string, ChapterImageLoad>()

  private constructor(
    bookId: string,
    crawler: WebCrawler,
    version: BookVersion,
    legacyImportGenerationKey: string,
    requestControlFactory?: CrawlerRequestControlFactory,
    resources?: BookResourceCache,
  ) {
    this.bookId = bookId
    this.crawler = crawler
    this.version = version
    this.legacyImportGenerationKey = legacyImportGenerationKey
    this.requestControlFactory = requestControlFactory
    this.resources = resources
  }

  get generationKey(): string {
    return this.version.generationKey
  }

  get versionFields(): BookVersionFields {
    return { ...this.version.fields }
  }

  static async create(
    bookId: string,
    crawler: WebCrawler,
    signal?: AbortSignal,
    requestControlFactory?: CrawlerRequestControlFactory,
    resources?: BookResourceCache,
  ): Promise<Book> {
    const page = await Book.fetchPage(bookId, crawler, signal, requestControlFactory)
    const version = createBookVersion(page.versionFields, Date.now())
    return Book.createFromPage(
      bookId,
      crawler,
      page,
      version,
      version.generationKey,
      signal,
      requestControlFactory,
      resources,
    )
  }

  static async fetchPage(
    bookId: string,
    crawler: WebCrawler,
    signal?: AbortSignal,
    requestControlFactory?: CrawlerRequestControlFactory,
  ): Promise<ParsedBookPage> {
    const version = createBookVersion({ updatedAt: '', latestChapter: '', status: '' }, 0)
    const parser = new Book(bookId, crawler, version, version.generationKey, requestControlFactory)
    const bookUrl = `https://www.wenku8.net/book/${bookId}.htm`
    const control = parser.requestControl('document', bookUrl)
    const $ = control
      ? await crawler.fetch(bookUrl, true, signal, control)
      : await crawler.fetch(bookUrl, true, signal)

    let rawChapterIndexUrl = ''
    $('#content div a').each((_i, element) => {
      const link = $(element)
      if (link.text().includes('小说目录') && link.attr('href')) {
        rawChapterIndexUrl = link.attr('href')!
        return false
      }
    })
    if (!rawChapterIndexUrl) throw new Error('未找到小说目录链接')

    const contentDiv = $('#content')
    const table = contentDiv.find('table').first()
    const title = table.find('b').first().text().trim()
    const cells: string[] = []
    table.find('tr').eq(2).find('td').each((_i, td) => cells.push($(td).text().trim()))

    let latestChapter: string | null = null
    let description = ''
    let tags: string[] = []
    let animated = false
    let heat: string | null = null
    $('#content span.hottext').each((_i, element) => {
      const text = $(element).text().replace(/\s+/g, ' ').trim()
      if (/\bTags\s*[:：]/i.test(text)) {
        const values = text.split(/\bTags\s*[:：]/i).slice(1).join(':')
        const seen = new Set<string>()
        tags = values
          .split(/[\s,，、]+/)
          .map(value => value.trim())
          .filter(value => {
            if (!value || value.length > 50 || seen.has(value)) return false
            seen.add(value)
            return true
          })
          .slice(0, 30)
      }
      if (/已(?:动画|動畫)化/.test(text)) animated = true
      if (text.includes('作品热度') || text.includes('作品熱度')) {
        heat = text.split(/(?:作品热度|作品熱度)\s*[:：]/).slice(1).join(':').trim() || null
      }
      if (text.includes('最新章节') || text.includes('最近章节')) {
        const link = $(element).nextAll('span').first().find('a')
        if (link.length > 0) latestChapter = link.text().trim()
      }
      if (text.includes('内容简介')) {
        description = $(element).nextAll('span').first().text().trim()
      }
    })

    const basicInfo: BasicInfo = {
      '标题': title,
      '作者': cells[1]?.split('：')[1] || '',
      '出版社': cells[0]?.split('：')[1] || '',
      '最新章节': latestChapter,
      '连载状态': cells[2]?.split('：')[1] || '',
      '更新时间': cells[3]?.split('：')[1] || null,
      '全文长度': cells[4]?.split('：')[1] || null,
      '简介': description,
      '标签': tags,
      '动画化': animated,
      '热度': heat,
      'cover': contentDiv.find('img').first().attr('src') || null,
    }
    return {
      basicInfo,
      chapterIndexUrl: new URL(rawChapterIndexUrl, bookUrl).toString(),
      versionFields: {
        updatedAt: basicInfo['更新时间'] ?? '',
        latestChapter: basicInfo['最新章节'] ?? '',
        status: basicInfo['连载状态'],
      },
    }
  }

  static async createFromPage(
    bookId: string,
    crawler: WebCrawler,
    page: ParsedBookPage,
    version: BookVersion,
    legacyImportGenerationKey: string,
    signal?: AbortSignal,
    requestControlFactory?: CrawlerRequestControlFactory,
    resources?: BookResourceCache,
  ): Promise<Book> {
    const book = new Book(
      bookId,
      crawler,
      version,
      legacyImportGenerationKey,
      requestControlFactory,
      resources,
    )
    const control = book.requestControl('document', page.chapterIndexUrl)
    const $ = control
      ? await crawler.fetch(page.chapterIndexUrl, true, signal, control)
      : await crawler.fetch(page.chapterIndexUrl, true, signal)
    const volumes: Record<string, Chapter[]> = {}
    let currentVolume = ''
    $('table.css tr').each((_i, tr) => {
      const volumeCell = $(tr).find('td.vcss')
      if (volumeCell.length > 0) {
        currentVolume = volumeCell.text().trim()
        volumes[currentVolume] = []
        return
      }
      $(tr).find('a').each((_j, anchor) => {
        const name = $(anchor).text().trim()
        const link = $(anchor).attr('href')
        if (currentVolume && name && link) volumes[currentVolume].push({ name, link })
      })
    })
    book.baseChapterUrl = new URL('.', page.chapterIndexUrl).toString()
    book.volumes = volumes
    book.pictureUrls = book.buildPictureUrlMap()
    book.basicInfo = { ...page.basicInfo, '标签': [...page.basicInfo['标签']] }
    return book
  }

  static fromSnapshot(
    snapshot: BookSnapshot,
    crawler: WebCrawler,
    requestControlFactory?: CrawlerRequestControlFactory,
    resources?: BookResourceCache,
  ): Book {
    const book = new Book(
      snapshot.bookId,
      crawler,
      snapshot.version,
      snapshot.legacyImportGenerationKey,
      requestControlFactory,
      resources,
    )
    book.baseChapterUrl = snapshot.baseChapterUrl
    book.volumes = Object.fromEntries(
      Object.entries(snapshot.volumes).map(([name, chapters]) => [
        name,
        chapters.map(chapter => ({ ...chapter })),
      ]),
    )
    book.pictureUrls = book.buildPictureUrlMap()
    book.basicInfo = { ...snapshot.basicInfo, '标签': [...snapshot.basicInfo['标签']] }
    return book
  }

  toSnapshot(checkedAt: number): BookSnapshot {
    return {
      schemaVersion: 2,
      bookId: this.bookId,
      checkedAt,
      version: {
        fields: { ...this.version.fields },
        generationKey: this.version.generationKey,
        stable: this.version.stable,
      },
      legacyImportGenerationKey: this.legacyImportGenerationKey,
      baseChapterUrl: this.baseChapterUrl,
      volumes: Object.fromEntries(
        Object.entries(this.volumes).map(([name, chapters]) => [
          name,
          chapters.map(chapter => ({ ...chapter })),
        ]),
      ),
      basicInfo: { ...this.basicInfo, '标签': [...this.basicInfo['标签']] },
    }
  }

  private requestControl(
    kind: CrawlerRequestKind,
    url: string,
    requestControlFactory = this.requestControlFactory,
  ) {
    if (!requestControlFactory) return undefined
    let absoluteUrl = url
    try {
      absoluteUrl = new URL(url, 'https://www.wenku8.net/').toString()
    } catch {
      // WebCrawler surfaces malformed URLs; the scheduler still receives a bounded key.
    }
    return requestControlFactory(kind, absoluteUrl)
  }

  private buildPictureUrlMap(): Record<string, string> {
    const map: Record<string, string> = {}
    for (const [volumeName, chapters] of Object.entries(this.volumes)) {
      const illustration = chapters.find(chapter => chapter.name === '插图')
      if (illustration) map[volumeName] = illustration.link
    }
    return map
  }

  async getChapterImageUrls(
    volumeName?: string,
    signal?: AbortSignal,
    requestControlFactory?: CrawlerRequestControlFactory,
  ): Promise<string[] | null> {
    if (!volumeName) return null
    throwIfDownloadCancelled(signal)
    if (this.chapterImageUrlCache.has(volumeName)) {
      return this.cloneUrls(this.chapterImageUrlCache.get(volumeName) ?? null)
    }
    const pictureUrl = this.pictureUrls[volumeName]
    if (!pictureUrl) return null
    const url = new URL(pictureUrl, this.baseChapterUrl).toString()
    let entry = this.chapterImageUrlLoads.get(volumeName)
    if (entry?.controller.signal.aborted) {
      this.chapterImageUrlLoads.delete(volumeName)
      entry = undefined
    }
    if (!entry) {
      const controller = new AbortController()
      const promise = this.loadChapterImageUrls(url, controller.signal, requestControlFactory)
      const created: ChapterImageLoad = { controller, promise, waiters: new Set() }
      entry = created
      this.chapterImageUrlLoads.set(volumeName, created)
      promise.then(
        (urls) => {
          if (this.chapterImageUrlLoads.get(volumeName) !== created) return
          this.chapterImageUrlLoads.delete(volumeName)
          if (created.controller.signal.aborted) return
          this.chapterImageUrlCache.set(volumeName, this.cloneUrls(urls))
        },
        () => {
          if (this.chapterImageUrlLoads.get(volumeName) === created) {
            this.chapterImageUrlLoads.delete(volumeName)
          }
        },
      )
    }
    return this.cloneUrls(await waitForChapterImages(entry, signal))
  }

  private async loadChapterImageUrls(
    url: string,
    signal?: AbortSignal,
    requestControlFactory?: CrawlerRequestControlFactory,
  ): Promise<string[] | null> {
    const guard = this.resources?.captureResourceWriteGuard(this.bookId, this.generationKey)
    const cached = await this.resources?.loadIllustration(this.bookId, this.generationKey, url)
    if (cached !== undefined) return cached
    const control = this.requestControl('document', url, requestControlFactory)
    const $ = control
      ? await this.crawler.fetch(url, true, signal, control)
      : await this.crawler.fetch(url, true, signal)
    const urls: string[] = []
    $('img').each((_i, image) => {
      const source = $(image).attr('src')
      if (!source) return
      try {
        urls.push(new URL(source, url).toString())
      } catch {
        // Ignore malformed illustration sources.
      }
    })
    const result = urls.length > 0 ? urls : null
    if (guard) {
      await this.resources?.saveIllustration(
        this.bookId,
        this.generationKey,
        url,
        result,
        guard,
      )
    }
    return result
  }

  async getCoverContent(
    signal?: AbortSignal,
    requestControlFactory?: CrawlerRequestControlFactory,
  ): Promise<Buffer> {
    const rawCoverUrl = this.basicInfo['cover']
    if (!rawCoverUrl) throw new Error('无封面图片')
    const coverUrl = new URL(rawCoverUrl, `https://www.wenku8.net/book/${this.bookId}.htm`).toString()
    const guard = this.resources?.captureResourceWriteGuard(this.bookId, this.generationKey)
    const cached = await this.resources?.loadCover(this.bookId, this.generationKey, coverUrl)
    if (cached) return Buffer.from(cached)
    const control = this.requestControl('image', coverUrl, requestControlFactory)
    const content = control
      ? await this.crawler.getImageContent(coverUrl, 3, undefined, signal, control)
      : await this.crawler.getImageContent(coverUrl, 3, undefined, signal)
    if (!content || content.byteLength === 0) throw new Error('封面下载失败')
    if (guard) {
      await this.resources?.saveCover(
        this.bookId,
        this.generationKey,
        coverUrl,
        content,
        guard,
      )
    }
    return content
  }

  private cloneUrls(urls: string[] | null): string[] | null {
    return urls ? [...urls] : null
  }

  getFormattedTitle(format: TitleFormat): string {
    return formatBookTitle(this.basicInfo['标题'], format)
  }
}
